depends/hosts/mingw32.mk

```
@@ -2,6 +2,7 @@ mingw32_CFLAGS=-pipe
 mingw32_CXXFLAGS=$(mingw32_CFLAGS) -isystem $(host_prefix)/include/c++/v1
 
 mingw32_LDFLAGS?=-fuse-ld=lld
+mingw32_LDFLAGS+=-L/usr/lib/gcc/x86_64-w64-mingw32/$(shell x86_64-w64-mingw32-gcc -dumpversion)
```

This outputs `$(major_gcc_version)-$(threading_option)`, for example "10-posix".

`zcashd` uses `std::thread` APIs, which are [only available with the `posix` threading option](https://stackoverflow.com/a/30390278/393146). The other threading option, `win32`, cannot possibly work.

Note that we're actually building with `clang` and `clang++`. The right thing is to always use the posix variant of the libraries provided by the mingw packages (`g++-mingw-w64-x86-64-posix` on Debian). If we use `$(shell x86_64-w64-mingw32-g++-posix -dumpversion)`, I believe that will do the right thing, i.e. not fail in situations where it could possibly work.

Personally I always make sure that gcc and g++ are set to the posix variants:
```shell
sudo update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-posix
sudo update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix
```
although I'm not absolutely sure that's necessary.

```suggestion
mingw32_LDFLAGS+=-L/usr/lib/gcc/x86_64-w64-mingw32/$(shell x86_64-w64-mingw32-g++-posix -dumpversion)
```

---

> `zcashd` uses `std::thread` APIs, which are [only available with the `posix` threading option](https://stackoverflow.com/a/30390278/393146). The other threading option, `win32`, cannot possibly work.

That's exactly what I had been thinking before I came across https://github.com/zcash/zcash/issues/4848#issuecomment-780833893, where @str4d wrote 
>Note that the old instructions to use update-alternatives to configure the POSIX variant of x86_64-w64-mingw32-gcc and x86_64-w64-mingw32-g++ was not required; they remained in their win32 state.

In the same context, how this is possible to work also https://github.com/zcash/zcash/blob/342563eb48371cfaede83ded743224d24ae6eece/depends/packages/boost.mk#L20 

That note encouraged me to test both threading model alternatives (win32 and posix), and both builds have produced expected binaries successfully. Since default alternative config is set to auto and win32 gets picked because of higher priority, I've left it this way as it hasn't been the deal breaker.

Nevertheless, I'm not against proposed change, but would like to hear some thoughts from @str4d first, if that's not too much to ask.

---

As far as I understand, `std::thread` is provided by `libc++`, and its implementation isn't dependent on what alternatives `x86_64-w64-mingw32-g++` and `x86_64-w64-mingw32-gcc` are set to; in fact the latter *shouldn't* be used at all. (If we were using `libstdc++` with `g++`, this support would be determined by the thread model, and in that case it would only work when using the `posix` variant.)

We can verify that setting these alternatives doesn't affect the threading model used by clang++. On my system, the latter is always `posix`:
```shell
$ sudo update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix
update-alternatives: using /usr/bin/x86_64-w64-mingw32-g++-posix to provide /usr/bin/x86_64-w64-mingw32-g++ (x86_64-w64-mingw32-g++) in manual mode
$ sudo update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-posix
update-alternatives: using /usr/bin/x86_64-w64-mingw32-gcc-posix to provide /usr/bin/x86_64-w64-mingw32-gcc (x86_64-w64-mingw32-gcc) in manual mode
$ depends/x86_64-w64-mingw32/native/bin/clang++ -v
clang version 15.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
[...]
$ sudo update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-win32
update-alternatives: using /usr/bin/x86_64-w64-mingw32-g++-win32 to provide /usr/bin/x86_64-w64-mingw32-g++ (x86_64-w64-mingw32-g++) in manual mode
daira@katava:~/olddaira/zecc/zcash.pristine$ sudo update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-win32
update-alternatives: using /usr/bin/x86_64-w64-mingw32-gcc-win32 to provide /usr/bin/x86_64-w64-mingw32-gcc (x86_64-w64-mingw32-gcc) in manual mode
$ depends/x86_64-w64-mingw32/native/bin/clang++ -v
clang version 15.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
[...]
```

However, we do also use `libgcc` (not `libunwind` and `compiler-rt`), and that is different depending on which library path under `/usr/lib/gcc/x86_64-w64-mingw32/` we specify:
```shell
$ ls -l /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc*
-rw-r--r-- 1 root root 2799650 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc.a
-rw-r--r-- 1 root root    9940 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc_eh.a
-rw-r--r-- 1 root root   68070 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc_s.a
-rwxr-xr-x 1 root root  609852 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc_s_seh-1.dll
$ ls -l /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc*
-rw-r--r-- 1 root root 2803762 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc.a
-rw-r--r-- 1 root root   10826 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc_eh.a
-rw-r--r-- 1 root root   68070 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc_s.a
-rwxr-xr-x 1 root root  628507 Apr 15  2022 /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc_s_seh-1.dll
```
`libc++` implements both threading models on Windows, [as described here](https://libcxx.llvm.org/DesignDocs/ThreadingSupportAPI.html). (The macros there are only used when compiling `libc++`.)

If we use the `win32` variants of `libgcc*`, it's likely that is a less well tested configuration, for two reasons:
* As seen above, `clang++` is always configured to use the `posix` model. I couldn't find documentation on what difference this makes, but presumably there is a difference otherwise it wouldn't be printed by `clang++ -v`.
* I think the `win32` variants of `libgcc*` won't have been tested as thoroughly with C++ standard library threads because `libstdc++` doesn't support that (and so that option will only have been tested by people using `libc++`).
* It's almost certainly a bad idea to mix models between the one that `clang++` thinks it's using and the one that is used for the library path.

There is a case for zcashd to switch to using `libunwind` and `compiler-rt`, but I won't complicate things by discussing that here, and it shouldn't be necessary to fix the Windows cross-build.

---

Aha:
```shell
$ objdump --syms /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc.a >libgcc-posix.syms
$ objdump --syms /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc.a >libgcc-win32.syms
$ diff libgcc-posix.syms libgcc-win32.syms 
1c1
< In archive /usr/lib/gcc/x86_64-w64-mingw32/10-posix/libgcc.a:
---
> In archive /usr/lib/gcc/x86_64-w64-mingw32/10-win32/libgcc.a:
3684a3685,3723
> 
> 
> 
> gthr-win32.o:     file format pe-x86-64
> 
> SYMBOL TABLE:
> [  0](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 1) 0x0000000000000000 __gthr_win32_once
> AUX tagndx 0 ttlsiz 0x0 lnnos 0 next 0
> [  2](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000080 __gthr_win32_key_create
> [  3](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x00000000000000c0 __gthr_win32_key_delete
> [  4](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x00000000000000f0 __gthr_win32_getspecific
> [  5](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000130 __gthr_win32_setspecific
> [  6](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000160 __gthr_win32_mutex_init_function
> [  7](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000190 __gthr_win32_mutex_destroy
> [  8](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x00000000000001a0 __gthr_win32_mutex_lock
> [  9](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x00000000000001f0 __gthr_win32_mutex_trylock
> [ 10](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000210 __gthr_win32_mutex_unlock
> [ 11](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000250 __gthr_win32_recursive_mutex_init_function
> [ 12](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000290 __gthr_win32_recursive_mutex_lock
> [ 13](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000320 __gthr_win32_recursive_mutex_trylock
> [ 14](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000370 __gthr_win32_recursive_mutex_unlock
> [ 15](sec  1)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x00000000000003c0 __gthr_win32_recursive_mutex_destroy
> [ 16](sec  1)(fl 0x00)(ty    0)(scl   3) (nx 1) 0x0000000000000000 .text
> AUX scnlen 0x3d5 nreloc 21 nlnno 0
> [ 18](sec  4)(fl 0x00)(ty    0)(scl   3) (nx 1) 0x0000000000000000 .xdata
> AUX scnlen 0x80 nreloc 0 nlnno 0
> [ 20](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_Sleep
> [ 21](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_TlsAlloc
> [ 22](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_GetLastError
> [ 23](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_TlsFree
> [ 24](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_TlsGetValue
> [ 25](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_SetLastError
> [ 26](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_TlsSetValue
> [ 27](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_CreateSemaphoreW
> [ 28](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_CloseHandle
> [ 29](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_WaitForSingleObject
> [ 30](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_ReleaseSemaphore
> [ 31](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_GetCurrentThreadId
> [ 32](sec  0)(fl 0x00)(ty   20)(scl   2) (nx 0) 0x0000000000000000 __mingwthr_key_dtor
```
In other words, the difference between `10-posix/libgcc.a` and `10-win32/libgcc.a` is that the latter includes an additional object file `gthr-win32.o` that exports the symbols above. Neither `libc++` nor `libc++abi` use these symbols:
```shell
$ objdump --syms depends/x86_64-w64-mingw32/native/lib/x86_64-unknown-linux-gnu/libc++.a |grep __gthr_win32
$ objdump --syms depends/x86_64-w64-mingw32/native/lib/x86_64-unknown-linux-gnu/libc++abi.a |grep __gthr_win32
$ objdump --syms depends/x86_64-w64-mingw32/native/lib/x86_64-unknown-linux-gnu/libc++.a |grep __imp_
$ objdump --syms depends/x86_64-w64-mingw32/native/lib/x86_64-unknown-linux-gnu/libc++abi.a |grep __imp_
```
Which explains why it worked for you. But this is a detail of the current implementation, and mixing thread models is still a bad idea.

As for Boost:
> In the same context (threading model), how this is possible to work also? https://github.com/zcash/zcash/blob/342563eb48371cfaede83ded743224d24ae6eece/depends/packages/boost.mk#L20 
This option only affects how Boost threads are implemented. It is probably a very bad idea for us to be using both `std::thread` and `boost::thread`, but that's true regardless of which threading model each uses. Note that Bitcoin Core no longer uses Boost threads; it exclusively uses `std::thread`. This appears to have been completed by Bitcoin Core v22.0.